src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.NormalSearch.cs

```diff
@@ -2,92 +2,117 @@
             Document document,
             string searchPattern,
             IImmutableSet<string> kinds,
             Document? activeDocument,
-            Func<INavigateToSearchResult, Task> onResultFound,
-            CancellationToken cancellationToken)
+            [EnumeratorCancellation] CancellationToken cancellationToken)
         {
-            var solution = document.Project.Solution;
-            var onItemFound = GetOnItemFoundCallback(solution, activeDocument, onResultFound, cancellationToken);
-
             var client = await RemoteHostClient.TryGetClientAsync(document.Project, cancellationToken).ConfigureAwait(false);
-            if (client != null)
+            await foreach (var item in ConvertItemsAsync(
+                document.Project.Solution, activeDocument, SearchDocumentWorkerAsync(), cancellationToken).ConfigureAwait(false))
             {
-                var callback = new NavigateToSearchServiceCallback(onItemFound);
-                // Don't need to sync the full solution when searching a particular project.
-                await client.TryInvokeAsync<IRemoteNavigateToSearchService>(
-                    document.Project,
-                    (service, solutionInfo, callbackId, cancellationToken) =>
-                    service.SearchDocumentAsync(solutionInfo, document.Id, searchPattern, kinds.ToImmutableArray(), callbackId, cancellationToken),
-                    callback, cancellationToken).ConfigureAwait(false);
-
-                return;
+                yield return item;
             }
 
-            await SearchDocumentInCurrentProcessAsync(document, searchPattern, kinds, onItemFound, cancellationToken).ConfigureAwait(false);
+            IAsyncEnumerable<RoslynNavigateToItem> SearchDocumentWorkerAsync()
+            {
+                if (client != null)
+                {
+                    var channel = Channel.CreateUnbounded<RoslynNavigateToItem>();
+
+                    // Kick off the work to do the search in another thread.  That work will push the results into the
+                    // channel.  When the work finishes (for any reason, including cancellation), the channel will be 
+                    // completed.
+                    Task.Run(async () => await client.TryInvokeAsync<IRemoteNavigateToSearchService>(
+                            // Don't need to sync the full solution when searching a particular project.
+                            document.Project,
+                            (service, solutionInfo, callbackId, cancellationToken) =>
+                                service.SearchDocumentAsync(solutionInfo, document.Id, searchPattern, kinds.ToImmutableArray(), callbackId, cancellationToken),
+                            new NavigateToSearchServiceCallback(channel), cancellationToken).ConfigureAwait(false), cancellationToken)
+                        .CompletesChannel(channel);
+
+                    return channel.Reader.ReadAllAsync(cancellationToken);
```

Comment:
worth having a helper that wraps the channel bits?  Since its repeated a few times

Reply:
i kept trying.  but i couldn't find anything i liked :)